### PR TITLE
Database upgrade script

### PR DESCRIPTION
**Pull request by Stell0**
https://github.com/nethesis/dev/issues/5791

### CODE REVIEW

File: src/Upgrade.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    $current_version = getVersion();
+    $next_version = $current_version + 1;
+    $container['logger']->info("Upgrading from $current_version to $next_version...");
+
+    # Launch update scripts
+    if (class_exists("\\Tancredi\\Upgrades\\Upgrader" . $next_version)) {
+        $upgrader = "\\Tancredi\\Upgrades\\Upgrader" . $next_version;
+        $upgraderObj = new $upgrader($container['storage'], $container['config'], $container['logger']);
+        $results = $upgraderObj();
+    }
```
Comment (DavidePrincipi) [inline src/Upgrade.php:36]: I wouldn't use OOP here. A plain PHP `include` statement could do the job...

Comment (DavidePrincipi) [inline src/Upgrade.php:36]: I wouldn't use OOP here. A plain PHP include statement could do the job...

File: src/Upgrade.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    # Increment current version
+    break;
+    $scope = new \Tancredi\Entity\Scope('defaults', $container['storage'], $container['logger']);
+    $scope->metadata['version'] = $next_version ;
+    $scope->setVariables();
+}
+
+
+function getVersion($current_version = true) {
```
Comment (DavidePrincipi) [inline src/Upgrade.php:46]: The argument of this function is never passed. What is it for?

File: src/Upgrade.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    global $container;
+    $defaults = $container['storage']->storageRead('defaults', $original = !$current_version);
```
[the remark this answers is not in the document]
Comment (Stell0) [reply src/Upgrade.php:47]: it's the argument for storageRead https://github.com/nethesis/tancredi/blob/7b0b935b94e1bef853c9cb198311a0e5cc42259f/src/Entity/FileStorage.php#L13 , if true read from read only directory

File: src/Upgrade.php (hunk 2 above)
Comment (Stell0) [reply src/Upgrade.php:45]: It is passed here https://github.com/nethesis/tancredi/blob/8dabbd2c5f6d4ab9eefe7ca78579696cb8938eca/src/Upgrade.php#L24 to get the starting version (the one in read write directory)
Maybe code isn't clear here, I'm open for suggestions

File: src/Upgrade.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    global $config;
+    $storage = new \Tancredi\Entity\FileStorage($c['logger'],$config);
+    return $storage;
+};
+
+$container['logger']->info("Launching upgrade scripts");
+# Launch update scripts
+$filesArray=glob(__DIR__ . "/upgrades.d/*.php");
+foreach ($filesArray as $file) {
+    $container['logger']->debug("Launching upgrade script $file");
```
Comment (DavidePrincipi) [inline src/Upgrade.php:22]:
```suggestion
    $container['logger']->info("Launching upgrade script $file");
```

File: src/Upgrade.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    $incf = function() use ($container,$file) {
+        include $file;
+    };
+    $incf();
```
Comment (DavidePrincipi) [inline src/Upgrade.php:26]: Is a plain `include()` enough?

```suggestion
    include $file;
```

File: src/Upgrade.php [hunk 6]
```diff
@@ -0,0 +1,28 @@
+<?php namespace Tancredi;
+
+require_once '../vendor/autoload.php';
+use Pimple\Container;
+
+$container = new Container();
```
Comment (DavidePrincipi) [inline src/Upgrade.php:6]:
```suggestion
$container = new \Pimple\Container();
```

File: src/Upgrade.php [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+$container['storage'] = function($c) {
+    global $config;
+    $storage = new \Tancredi\Entity\FileStorage($c['logger'],$config);
+    return $storage;
+};
+
+$container['logger']->info("Launching upgrade scripts");
+# Launch update scripts
+$filesArray=glob(__DIR__ . "/upgrades.d/*.php");
```
Comment (DavidePrincipi) [inline src/Upgrade.php:20]:
```suggestion
$filesArray=glob(__DIR__ . "/upgrade.d/*.php");
```

File: src/Upgrade.php [hunk 8]
```diff
@@ -0,0 +1,28 @@
+<?php namespace Tancredi;
+
+require_once '../vendor/autoload.php';
+use Pimple\Container;
+
+$container = new Container();
+$container['config'] = $config;
+$container['logger'] = function($c) {
+    return \Tancredi\LoggerFactory::createLogger('Upgrader', $c);
```
Comment (DavidePrincipi) [inline src/Upgrade.php:9]:
```suggestion
    return \Tancredi\LoggerFactory::createLogger('upgrade', $c);
```

File: src/upgrades.d/00001.php [hunk 9]
```diff
@@ -0,0 +1,56 @@
+<?php
+
+namespace upgrade1;
+
+$current_version = getDefaultsVersion();
```
Comment (DavidePrincipi) [inline src/upgrades.d/00001.php:5]: We could move it in the general script: it looks to be useful in the future...

File: src/upgrades.d/00001.php [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+if ($current_version === 0 ) {
```
Comment (DavidePrincipi) [inline src/upgrades.d/00001.php:6]: Exit early

```suggestion
if ($current_version != 0 ) {
    return;
}
```

File: src/upgrades.d/00001.php [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        $scope = new \Tancredi\Entity\Scope($id, $container['storage'], $container['logger']);
+        if (array_key_exists($scope->metadata['inheritFrom'], $names)) {
+            $scope->metadata['inheritFrom'] = $names[$scope->metadata['inheritFrom']];
+            $scope->setVariables();
+        }
+    }
+
+    # Increment defaults version
+    $scope = new \Tancredi\Entity\Scope('defaults', $container['storage'], $container['logger']);
+    $scope->metadata['version'] = $current_version + 1 ;
```
Comment (DavidePrincipi) [inline src/upgrades.d/00001.php:43]: The version must be constant!

```suggestion
    $scope->metadata['version'] = 1 ;
```

File: src/upgrades.d/00001.php (hunk 9 above)
Comment (Stell0) [reply src/upgrades.d/00001.php]: can I use it if it is in a different namespace?

Comment (DavidePrincipi) [reply src/upgrades.d/00001.php]: It's required a namespace prefix in that case, or to import it with "use"

File: src/upgrade.d/001.php [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+
+function getDefaultsVersion($original = false) {
+    global $container;
+    $defaults = $container['storage']->storageRead('defaults', $original);
+    $version = 0;
+    if (array_key_exists('metadata',$defaults) && array_key_exists('version',$defaults['metadata'])) {
+        $version = $defaults['metadata']['version'];
+    }
+    return $version;
```
Comment (DavidePrincipi) [inline src/upgrade.d/001.php:56]:
```suggestion
    if (empty($defaults['metadata']['version'])) {
        return 0;
    }
    return intval($defaults['metadata']['version']);
```

File: src/Upgrade.php [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+    return \Tancredi\LoggerFactory::createLogger('upgrade', $c);
+};
+
+$container['storage'] = function($c) {
+    global $config;
+    $storage = new \Tancredi\Entity\FileStorage($c['logger'],$config);
+    return $storage;
+};
+
+$container['logger']->info("Launching upgrade scripts");
```
Comment (DavidePrincipi) [inline src/Upgrade.php:16]: This log message is redundant

```suggestion
```

File: src/upgrade.d/001.php [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+    "yealink-T57" => "Yealink T57W",
+    "yealink-T58" => "Yealink T58V/A"
+);
+
+# Rename scope files
+foreach ($displaynames as $id => $new_displayname) {
+    $scope = new \Tancredi\Entity\Scope($id, $container['storage'], $container['logger']);
+    if (empty($scope->metadata['version']) && $scope->metadata['displayName'] != $new_displayname) {
+        $scope->metadata['displayName'] = $new_displayname;
+        $scope->metadata['version'] = 1;
```
Comment (DavidePrincipi) [inline src/upgrade.d/001.php:67]: Commit changes:

```suggestion
        $scope->metadata['version'] = 1;
        $scope->setVariables();
```

**Pull request merged**